lib/storage/ldp/ldp-file-store.js

```diff
@@ -0,0 +1,56 @@
+'use strict'
+
+const fs = require('fs-extra')
+const LdpFileResource = require('./ldp-file-resource')
+
+class LdpFileStore {
+  /**
+   * @param options {object}
+   *
+   * @param options.root {string} Root fs path for data storage
+   * @param options.multiuser {boolean}
+   * @param options.suffixAcl {string}
+   * @param options.suffixMeta {string}
+   * @param options.mapper {ResourceMapper}
+   * @param options.dataBrowserPath {string}
+   * @param options.suppressDataBrowser
+   */
+  constructor (options) {
+    this.root = options.root
+    this.multiuser = options.multiuser
+    this.suffixAcl = options.suffixAcl
+    this.suffixMeta = options.suffixMeta
+
+    this.mapper = options.mapper
+
+    this.dataBrowserPath = options.dataBrowserPath
+    this.suppressDataBrowser = options.suppressDataBrowser
+  }
+
+  /**
+   * @param target {LdpTarget}
+   *
+   * @returns {Promise<LdpFileResource>}
+   */
+  async resource (target) {
+    const filePath = await this.mapper
+      .mapUrlToFile({
+        url: target.url, contentType: target.mediaType(), createIfNotExists: true
```

Comment:
@RubenVerborgh Ok, so now we're at the stage of sketching the backend / LdpFileStore.
I suspect we'll need to refactor either ResourceMapper or LegacyResourceMapper.

The current behavior of `mapUrlToFile()` is to either 1) create the filepath using $ (non-legacy behavior) if `createIfNotExists: true`, or 2) Throw a 'File not found' error if `createIfNotExists: false` (which is always true if called via `LegacyResourceMapper`).

I'd like a slightly different semantics - to *not* throw a 'file not found' error, and just return "if the file existed this is what its path would be" in legacy mode.

Reply:
I hope no major refactoring of `LegacyResourceMapper` is necessary; I tried to design it so it can be plugged in right away.

Let's ignore the new `ResourceMapper` for now and let's check the legacy one, which intends to capture the current behavior in Solid.

> 2) Throw a 'File not found' error if createIfNotExists: false (which is always true if called via LegacyResourceMapper).

You can't run into such an error with `LegacyResourceMapper`. It is a simple functional mapping between a URL and a filename, which, unlike `ResourceMapper`, does not depend on the filesystem. The legacy mapper (aka current behavior) is independent of whether the file already exists, it's a simple 1-on-1 mapping, regardless of disk contents. This changes with the `ResourceMapper`, which follows https://www.w3.org/DesignIssues/HTTPFilenameMapping.html instead.

> I'd like a slightly different semantics - to not throw a 'file not found' error, and just return "if the file existed this is what its path would be" in legacy mode.

We should have that already. If not, you probably found a bug.

Reply:
> You can't run into such an error with `LegacyResourceMapper`

Ohhh I see it now. I misread. n/m.